### PR TITLE
Angelica Ceja - API Muncher - Octos

### PR DESCRIPTION
**Pull request by amcejamorales**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I used Postman to try out different search test cases. I then took the URL that worked on Postman and used it in my code. I also used Postman to troubleshoot/test different query parameters. 
Describe your API Wrapper.  How did you decide on the methods you created? | My API Wrapper has two methods: list_recipes and show_recipe. List_recipe takes in one argument (a search term which I take from the search form in the view), it then uses HTTParty (which takes in a curated URL using the search term) to get data from the API. I iterate over this data and call a private method, create_recipe, to create a new instance of the class Recipe for each data point. Finally, I return an array containing these instances of Recipe. Show_recipe takes in a uri id, curates the full search url using this id, gets data using HTTParty and creates an instance of Recipe with this data (this method also calls create_recipe). I also have other private methods in the Wrapper to get the uri_id from the raw data taken from the Edamam API and to format the search uri for a single recipe. 
Describe an edge case or failure case test you wrote for your API Wrapper. | I tested for both a blank search and a bogus search and verified that my list_recipes method returns an empty array in both cases. The bogus search is long string of x's so it is safe to assume this would be unlikely to show up in a recipe in the Edamam API. 
Explain how VCR aids in testing an API. | It created 'cassettes', which hold test data for us from the first time the tests were run. This saves us trying to call the API each time we try to test in the future. 
What is the Heroku URL of your deployed application? | https://comelon-ac.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/invite/b/gclyp3TS/e2370b9749f208c4f7899b2133bd3537/muncher

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    <%= csrf_meta_tags %>
+    <script src="https://developer.edamam.com/attribution/badge.js"></script>
+  </head>
+
+  <body>
+
+    <section>
+      <h1><%= link_to "Comelón", root_path %></h1>
+      <h4>¡Busca recetas con información dietética!</h4>
+      <%= form_tag recipes_path, method: :get do %>
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb:20]: Should this be a `<header>` instead of a `<section>`?

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      api_params['dietLabels'],
+      api_params['url'],
+      get_uri_id(api_params['uri'])
+    )
+  end # create_recipe
+
+  def self.get_uri_id(uri)
+    id = uri
+    if uri.include?("recipe_")
+      split_uri = uri.split("recipe_")
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:51]: Good use of helper methods throughout this file.

File: app/controllers/recipes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    search_term = params[:search]
+    recipes = EdamamApiWrapper.list_recipes(search_term)
+    if recipes
+      @recipes = Kaminari.paginate_array(recipes).page(params[:page]).per(MAX_RECIPES_PER_PAGE)
+    end
+  end
+
+  def show
+    @recipe = EdamamApiWrapper.show_recipe(params[:id])
+    head :bad_request unless @recipe
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:15]: You look for a `nil` recipe here, but in your lib file you raise an exception if the API didn't return a recipe. This is how I encountered that error by mangling the URI.

File: test/controllers/recipes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        get recipe_path(uri_id)
+
+        must_respond_with :ok
+      end
+    end # it do
+
+    it 'raises an error for an recipe that DNE' do
+      VCR.use_cassette("bad_recipe") do
+        uri_id = "x"
+        proc { get recipe_path(uri_id) }.must_raise
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:47]: In general, controller actions should never raise an exception - they should always handle the error and send back a reasonable status code / put an error message in the flash.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | yes - good work identifying that `resources` is still useful
Semantic HTML | yes
Errors are reported to the user | no - when I mangle the recipe URI, I get an exception
API Wrapper to handle the API requests | yes
Controller testing | yes
Lib testing | yes
Search Functionality | yes
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes (no popout link)
**Styling** | 
Responsive layout | yes
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | yes
**API Features** | 
The App attributes Edaman | yes
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | yes
**Overall** | Great job overall! This code is well-organized and easy to read, and it's clear the learning goals for this assignment were met. I've left a number of nitpick comments inline, but in general I am quite happy with this submission. Keep up the hard work!
